lib/promise/group.rb

```diff
@@ -0,0 +1,51 @@
+class Promise
+  class Group
+    attr_reader :promise
+
+    def initialize(inputs)
+      @promise = Promise.new
+      @inputs = inputs
+      @remaining = count_promises(inputs)
+      if @remaining.zero?
+        promise.fulfill(inputs)
+      else
+        chain_inputs(inputs)
+      end
+    end
+
+    private
+
+    def chain_inputs(inputs)
+      on_fulfill = method(:on_fulfill)
+      on_reject = promise.method(:reject)
+      each_promise(inputs) do |input_promise|
+        input_promise.then(on_fulfill, on_reject)
+      end
+    end
+
+    def on_fulfill(_result)
+      @remaining -= 1
+      if @remaining.zero?
+        result = @inputs.map { |obj| promise?(obj) ? obj.value : obj }
+        promise.fulfill(result)
+      end
+    end
+
+    def promise?(obj)
+      obj.instance_of?(Promise)
+    end
+
+    def count_promises(enumerable)
```

Comment:
this and `each_promise` don't need to take a param, they can just operate on `@inputs`; otherwise they belong as class methods